src/sources/file.rs

```diff
@@ -325,15 +322,25 @@ pub fn file_source(
     let multiline_config = config.multiline.clone();
     let message_start_indicator = config.message_start_indicator.clone();
     let multi_line_timeout = config.multi_line_timeout;
-    let checkpoints = checkpointer.view();
-    let finalizer = acknowledgements.then(|| {
+
+    let (finalizer, shutdown_checkpointer) = if acknowledgements {
+        let (finalizer, mut ack_stream) = OrderedFinalizer::<FinalizerEntry>::new(shutdown.clone());
+        let (send_shutdown, shutdown2) = oneshot::channel::<()>();
         let checkpoints = checkpointer.view();
-        OrderedFinalizer::new(shutdown.clone(), move |entry: FinalizerEntry| {
-            let checkpoints = Arc::clone(&checkpoints);
-            async move { checkpoints.update(entry.file_id, entry.offset) }
-        })
-    });
+        tokio::spawn(async move {
+            while let Some((status, entry)) = ack_stream.next().await {
+                if status == BatchStatus::Delivered {
+                    checkpoints.update(entry.file_id, entry.offset);
+                }
+            }
+            send_shutdown.send(())
+        });
+        (Some(finalizer), shutdown2.map(|_| ()).boxed())
```

Comment:
Documenting how this shutdown works / why it's rigged up this way would probably be helpful for someone finding this in the future.

Reply:
Agreed, will fix.